### ISSUE TITLE
Inaccurate writing of the individual letters

### ISSUE THREAD

**Issue by TeoColuccio** [PyFPDF/fpdf2#585]
I'm trying to bring my project written in js, entirely in python, and to do that I need to be able to write a single letter in a defined portion of the page.
For this I am using pdf.text(), but I also tried .cell() (after setting x and y with set_xy()).
The problem is that the location doesn't seem to be accurate.

The code is this:
```python
from fpdf import FPDF

pdf = FPDF()

stri = [
    {
        "coords": [443.563, 241.273],
        "text": "t"
    },
    {
        "coords": [446.332, 241.273],
        "text": "e"
    },
    {
        "coords": [450.756, 241.273],
        "text": "s"
    },
    {
        "coords": [454.631, 241.273],
        "text": "t"
    },
    {
        "coords": [457.401, 241.273],
        "text": "i"
    },
    {
        "coords": [460.171, 241.273],
        "text": "n"
    },
    {
        "coords": [465.152, 241.273],
        "text": "g"
    }]

pdf.add_page(format=[612, 792])
pdf.set_font("Times", size=20)

for char in stri:
    pdf.text(txt=char["text"], x=char["coords"][0], y=char["coords"][1])

pdf.set_xy(443.563, 261.273)
pdf.cell(txt="testing")

pdf.output("hello_world.pdf")
```

I printed "testing" once with the single letters and once entirely as a string.
I know it sounds strange, but in my project, using a js library (pdf-lib.js), those identical coordinates allow you to print "testing" exactly as if it were an entire string, while still printing letter by letter. Obviously using the same font and the same size.
I attach an image with the results:
![screen](https://user-images.githubusercontent.com/16000669/195893033-e4a50876-1cfb-4ab4-85ba-40beba810832.png)

**Comment by Lucas-C** [comment]
I agree with @TS40 answer (thanks for helping!)

I'm well aware about `fpdf2` arbitrary float rounding, and that's something I'm planning to give users more control on,
at some point in the future. If people are willing to contribute some MRs about this, I'd be happy to discuss the matter more and to welcome them.

**Comment by gmischler** [comment]
`.cell()` and `.multi_cell()` are designed to arrange text within a box, possibly with a printed border. To avoid overlap with that border, they add a small distance to the right and left of it, which result in a gap between the characters if you just continue after `new_x=XPos.RIGHT`.
Theoretically, you could use `.text()` to place each character individually, but to get that right you'd have to know the actual graphical width of each character, in order to get the location of the next one right (not sure how you arrived at the coordinates in your example, but they are evidently not correct).

The solution is be to use [`.write()`](https://pyfpdf.github.io/fpdf2/Text.html#write), which is specifically designed for the purpose of appending the next section of text exactly where the previous one ended.
```python
for c in "testing":
    pdf.write(c)
```

**Comment by Lucas-C** [comment]
Also, if you want to reuse the same coordinates you had in your `pdf-lib.js`-based project,
a solution could be to specify a custom `unit` parameter to [`FPDF`](https://pyfpdf.github.io/fpdf2/fpdf/fpdf.html#fpdf.fpdf.FPDF) constructor, in order to stretch all the coordinates at once.

With all those answers, does that solve your problem @TeoColuccio ?

**Comment by TeoColuccio** [comment]
> With all those answers, does that solve your problem @TeoColuccio ?

The best solution at the moment is to adjust the x coordinates as in your code:
```
for i, char in enumerate(stri):
    x = char["coords"][0] - i
    pdf.text(txt=char["text"], x=x, y=char["coords"][1])
```
But in fact, a definitive solution would be to round off the x and y variables better, perhaps with more decimal places.
If I'm able, I'll do a PR.

**Comment by gmischler** [comment]
> But in fact, a definitive solution would be to round off the x and y variables better, perhaps with more decimal places.

There are __*NO*__ rounding issues at play here.
If rounding was an issue at all, then it would generally not be possible to position text (or anything) correctly with fpdf2.

The coordinates in a PDF file are defined in points (1/72 inch ~= 0.35 mm).
At two decimals as in `f"{123.4567:.2f}"`, we specify the text position to an accuracy of 0.0035 mm. A deviation that small is imperceptible to the human eye, even when blown up to billboard size.
The output example given shows a visible deviation of around 1 mm, so that clearly must have a different reason.

Note that fpdf2 by default accepts coordinate input in mm.
Looking at the numbers in the example code, the difference between the x values is on average roughly 3-4 mm. This matches well with the graphical output (scaled down from a font size of 20 points), so fpdf2 appears to do exactly what it was told to do.
In other words: Your coordinates are in the wrong unit.

So now you have three workable solutions to chose from:
 1. Try the recommendation given in my previous answer, or
 1. as @Lucas-C suggested above, set the fpdf2 document units to whatever your coordinates actually are (looks like `points*2`), or
 1. convert your coordinates to mm

**Comment by Lucas-C** [comment]
> At two decimals as in f"{123.4567:.2f}", we specify the text position to an accuracy of 0.0035 mm. A deviation that small is imperceptible to the human eye, even when blown up to billboard size.

While I fully agree with your comment & analysis @gmischler,
I still supect that some `fpdf2` rounding issues combined with some PDF viewers own rendering logic
may be the reason why [Undying Dusk](https://lucas-c.itch.io/undying-dusk) sometimes presents minor visual glitchs
when several images are rendered very close to each other and pixel-perfect precision is required.

This may be the result of various numeric operations causing some minor floating-point arithmetic errors, like this:
```python
a = 1234.567; b = 45.67834; c = 0.0004
print( (a+b)+c == a+(b+c) )  # False
```
_cf._ https://en.wikipedia.org/wiki/Floating-point_arithmetic#Accuracy_problems

In the end, I think `fpdf2` could provide its users more control over the precision of the numbers inserted in PDF files,
if only so that a user could choose to reduce precision in order to produce slightly smaller PDFs.

**Comment by gmischler** [comment]
> I still supect that some `fpdf2` rounding issues combined with some PDF viewers own rendering logic
> may be the reason why [Undying Dusk](https://lucas-c.itch.io/undying-dusk) sometimes presents minor visual glitchs

Easy enough to find out, if you create the same page with more decimals in the coordinates, and check what those viewers make of it then.
If that should fix the glitches (which I doubt, see below), then offering an option to set the rounding precision might make sense, otherwise it seems like a waste of effort.

```python
a = 1234.567; b = 45.67834; c = 0.0004
print( (a+b)+c == a+(b+c) )  # False
```
Yes, digital arithmethics have to deal with such errors (we use `math.isclose()` a few times because of it), but that is not a problem we can solve by rounding further to the right ourselfes. Quite the opposite.
```python
>>> (a+b)+c
1280.2457399999998
>>> a+(b+c)
1280.24574
```
Here the rounding error happens at the very last (13th) decimal, which is typical for this kind of error. 
With our output, the comparison a viewer needs to make would rather be equivalent to this:

```python
print( round((a+b)+c, 2) == round(a+(b+c), 2) )  # True
```

So, by essentially cutting off any potential numerical instabilities, our rounding really helps to make sure that theoretically identical coordinates arrive as actually identical numerical values at the client. The more decimals we allow, the greater the chance for floating point inaccuracies to influence the output and potentially make such a comparison fail.

As to visual glitches, PDF viewers do sometimes make errors (compare #487), but there's little we can do to fix _their_ problems.

**Comment by TeoColuccio** [comment]
> * Try the recommendation given in my previous answer, or
> * as @Lucas-C suggested above, set the fpdf2 document units to whatever your coordinates actually are (looks like `points*2`), or
> * convert your coordinates to mm
You were absolutely right, unfortunately I hadn't paid attention to the size of the document which was expressed in pt and not in mm as the default for fpdf. By simply setting pt in the fpdf constructor I solved it.

**Comment by Lucas-C** [comment]
Glad you solved your issue @TeoColuccio !

**Comment by TeoColuccio** [comment]
> Glad you solved your issue @TeoColuccio !

Thank you very much!
